owltorrent-network/src/main/java/edu/rice/owltorrent/network/MeteringSystem.java

```diff
@@ -0,0 +1,194 @@
+package edu.rice.owltorrent.network;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import edu.rice.owltorrent.common.entity.TwentyByteId;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Metering system to keep track of peer, system, and piece-related metrics.
+ *
+ * @author bhaveshshah
+ */
+public class MeteringSystem {
+  public enum Metrics {
+    // Peer-related metrics
+    PEER_DOWNLOAD_TIME,
+
+    // System-related metrics
+    ENTIRE_DOWNLOAD_TIME,
+    EFFECTIVE_PIECES,
+    FAILED_PIECES,
+
+    // Piece-related metrics
+    PIECE_DOWNLOAD_TIME
+  }
+
+  // Maps to store metrics and associated values
+  public static ConcurrentHashMap<TwentyByteId, ConcurrentHashMap<Enum<?>, AtomicDouble>>
+      peerMetrics;
+  public static ConcurrentHashMap<Enum<?>, AtomicDouble> systemMetrics;
+  public static ConcurrentHashMap<Integer, ConcurrentHashMap<Enum<?>, AtomicDouble>> pieceMetrics;
+
+  public MeteringSystem() {
+    this.peerMetrics = new ConcurrentHashMap<>();
+    this.systemMetrics = new ConcurrentHashMap<>();
+    this.pieceMetrics = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Get metric for a specified peer.
+   *
+   * @param metricType type of metric (system, peer, piece)
+   * @param peerID peer id
+   * @param pieceIndex index of piece
+   * @param metricName metric to track
+   * @return metric value for peer
+   */
+  public AtomicDouble getMetric(
+      String metricType, TwentyByteId peerID, Integer pieceIndex, Enum<?> metricName) {
+    AtomicDouble defaultVal = new AtomicDouble(0);
+
+    if (metricType.equals("peer")) {
+      if (peerID == null || !peerMetrics.containsKey(peerID)) {
+        return defaultVal;
+      }
+      return peerMetrics.get(peerID).getOrDefault(metricName, defaultVal);
+    } else if (metricType.equals("system")) {
+      return systemMetrics.getOrDefault(metricName, defaultVal);
+    } else if (metricType.equals("piece")) {
+      if (pieceIndex == null || !pieceMetrics.containsKey(pieceIndex)) {
+        return defaultVal;
+      }
+      return pieceMetrics.get(pieceIndex).getOrDefault(metricName, defaultVal);
+    } else {
+      return defaultVal;
+    }
+  }
+
+  /**
+   * Add or update system-wide metric.
+   *
+   * @param metricName metric to track
+   * @param metricVal numerical value of metric
+   */
+  public void addSystemMetric(Enum<?> metricName, double metricVal) {
```

Comment:
metricName should be type Metrics instead of wildcard enum. (Same with other places)

Reply:
Fixed in next commit